## Installing required packages :

In [1]:
from sklearn.datasets import load_digits
from sklearn.metrics import f1_score
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn import datasets
from sklearn.metrics import precision_score
from sklearn.datasets import fetch_olivetti_faces
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_digits
from sklearn.datasets import fetch_olivetti_faces

import keras
from keras.models import Sequential
from sklearn.datasets import load_iris
from keras.metrics import Precision , Recall
from keras.utils import np_utils
from keras.layers import Dense, Conv2D, MaxPooling2D , Flatten, Dense, Dropout, Activation,  GlobalAveragePooling2D, ZeroPadding2D
from keras.models import Model
from keras.models import Sequential
from keras.layers import Dense
from keras import layers , models

import tensorflow as tf
from tensorflow.keras.utils import to_categorical, plot_model

import numpy as np
import matplotlib.pyplot as plt

## Loading datasets 
##### Iris dataset

In [2]:
iris = load_iris()

##### Digits dataset

In [3]:
digits = load_digits()

##### Olivetti dataset

In [4]:
faces = fetch_olivetti_faces()

downloading Olivetti faces from https://ndownloader.figshare.com/files/5976027 to /root/scikit_learn_data


In [5]:
X = digits.data
y = digits.target

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=64)
train_labels = to_categorical(y_train, num_classes=digits.target_names.shape[0])
test_labels = to_categorical(y_test, num_classes=digits.target_names.shape[0])

## Using only fully connected layers

In [7]:
model = Sequential()
model.add(Dense(16 , input_dim =X_train.shape[1], activation ='relu'))
model.add(Dense(8 , activation ='relu'))
model.add(Dense(digits.target_names.shape[0], activation ='softmax'))
model.compile(loss ='categorical_crossentropy', optimizer='adam', metrics =['accuracy',Precision(),Recall()])
model.fit (X_train, train_labels, epochs=20, batch_size=10)

Epoch 1/20
144/144 [==============================] - 1s 2ms/step - loss: 2.7179 - accuracy: 0.1802 - precision: 0.1042 - recall: 0.0104
Epoch 2/20
144/144 [==============================] - 0s 2ms/step - loss: 1.7520 - accuracy: 0.3834 - precision: 0.9435 - recall: 0.0814
Epoch 3/20
144/144 [==============================] - 0s 2ms/step - loss: 1.2887 - accuracy: 0.5449 - precision: 0.9214 - recall: 0.3264
Epoch 4/20
144/144 [==============================] - 0s 2ms/step - loss: 0.8670 - accuracy: 0.7035 - precision: 0.9029 - recall: 0.5498
Epoch 5/20
144/144 [==============================] - 0s 2ms/step - loss: 0.6359 - accuracy: 0.7836 - precision: 0.9205 - recall: 0.6771
Epoch 6/20
144/144 [==============================] - 0s 2ms/step - loss: 0.5209 - accuracy: 0.8163 - precision: 0.9186 - recall: 0.7383
Epoch 7/20
144/144 [==============================] - 0s 2ms/step - loss: 0.4555 - accuracy: 0.8504 - precision: 0.9267 - recall: 0.7655
Epoch 8/20
144/144 [=====================

In [8]:
print("Model evaluation on test data")
model.evaluate(X_test,test_labels)

Model evaluation on test data
12/12 [==============================] - 1s 3ms/step - loss: 0.1540 - accuracy: 0.9417 - precision: 0.9468 - recall: 0.9389


[0.15403658151626587,
 0.9416666626930237,
 0.9467787146568298,
 0.9388889074325562]

# Faces Dataset - Model with only fully connected layers

In [9]:
X = faces.data
y = faces.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=64)
train_labels = to_categorical(y_train, 40)
test_labels = to_categorical(y_test, 40)

In [10]:
model = Sequential()
model.add(Dense(16 , input_dim =X_train.shape[1], activation ='relu'))
model.add(Dense(32 , activation ='relu'))
model.add(Dense(64 , activation ='relu'))
model.add(Dense(40, activation ='softmax'))
model.compile(loss ='binary_crossentropy', optimizer='adam', metrics =['accuracy',Precision(),Recall()])
model.fit (X_train, train_labels, epochs=80, batch_size=16)

Epoch 1/80
20/20 [==============================] - 1s 4ms/step - loss: 0.5769 - accuracy: 0.0094 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 2/80
20/20 [==============================] - 0s 3ms/step - loss: 0.2085 - accuracy: 0.0344 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 3/80
20/20 [==============================] - 0s 2ms/step - loss: 0.1314 - accuracy: 0.0156 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 4/80
20/20 [==============================] - 0s 2ms/step - loss: 0.1218 - accuracy: 0.0312 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 5/80
20/20 [==============================] - 0s 3ms/step - loss: 0.1200 - accuracy: 0.0312 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 6/80
20/20 [==============================] - 0s 3ms/step - loss: 0.1195 - accuracy: 0.0312 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 7/80
20/20 [==============================] - 0s 3ms/step - loss: 0.1195 - accuracy: 0.0312 - precision_1: 0.000

In [11]:
print("Model evaluation on test data")
model.evaluate(X_test,test_labels)

Model evaluation on test data
3/3 [==============================] - 1s 5ms/step - loss: 0.1223 - accuracy: 0.0000e+00 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00


[0.12228582054376602, 0.0, 0.0, 0.0]

The evaluation results are so poor, as we expected it to be since this dataset has multiple classes (40) and the fully conncted layers can not take into consideration the spacial caracteristics like the Convolutional Neural Network does. Thats why we will try training a CNN on this dataset in order to have better results.

## Implementing VGG 16 - faces dataset

In [12]:
X = faces['images']
y = faces['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], X_train.shape[2], 1))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1], X_test.shape[2], 1))
train_labels = to_categorical(y_train, num_classes=40)
test_labels = to_categorical(y_test, num_classes=40)
X_train=np.dstack([X_train] * 3)
X_test=np.dstack([X_test]*3)
X_train = X_train.reshape(-1, 64,64,3)
X_test= X_test.reshape (-1,64,64,3)

vgg_16_model = Sequential()
vgg_16_model.add(Conv2D(input_shape=(64,64,3) ,filters=64,kernel_size=(3,3),padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=64,kernel_size=(3,3),padding="same", activation="relu"))
vgg_16_model.add(MaxPooling2D(pool_size=(2,2),strides=(2,2)))
vgg_16_model.add(Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(MaxPooling2D(pool_size=(2,2),strides=(2,2)))
vgg_16_model.add(Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(MaxPooling2D(pool_size=(2,2),strides=(2,2)))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(MaxPooling2D(pool_size=(2,2),strides=(2,2)))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"))
vgg_16_model.add(MaxPooling2D(pool_size=(2,2),strides=(2,2)))
vgg_16_model.add(Flatten())
vgg_16_model.add(Dense(units=4096,activation="relu"))
vgg_16_model.add(Dense(units=4096,activation="relu"))
vgg_16_model.add(Dense(units=40, activation="softmax"))

In [13]:
X_test.shape

(80, 64, 64, 3)

In [14]:
vgg_16_model.compile(loss ='categorical_crossentropy', optimizer='adam', metrics =['accuracy',Precision(),Recall()])
vgg_16_model.fit (X_train, train_labels, epochs=15)

Epoch 1/15
10/10 [==============================] - 56s 5s/step - loss: 3.7754 - accuracy: 0.0219 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 2/15
10/10 [==============================] - 55s 6s/step - loss: 3.6897 - accuracy: 0.0125 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 3/15
10/10 [==============================] - 54s 5s/step - loss: 3.6886 - accuracy: 0.0250 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 4/15
10/10 [==============================] - 54s 5s/step - loss: 3.6856 - accuracy: 0.0312 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 5/15
10/10 [==============================] - 55s 5s/step - loss: 3.6832 - accuracy: 0.0312 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 6/15
10/10 [==============================] - 54s 5s/step - loss: 3.6845 - accuracy: 0.0156 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 7/15
10/10 [==============================] - 54s 5s/step - loss: 3.6811 - accuracy: 0.0250 - precision_2: 0.000

In [15]:
print("Model evaluation on test data")
vgg_16_model.evaluate(X_test,test_labels)

Model evaluation on test data
3/3 [==============================] - 5s 1s/step - loss: 3.7979 - accuracy: 0.0000e+00 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00


[3.797873020172119, 0.0, 0.0, 0.0]

VGG16 didn't give better results for faces dataset compared to the Fully connected neural network. We shall also add that it costs a lot of computational resources to train such a complex model. In order to get the benefits of this complex model without taking so much time and resources (using transfer learning).

### Transfer learning - VGG 16 - faces dataset

In [16]:
X = faces['images']
y = faces['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], X_train.shape[2], 1))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1], X_test.shape[2], 1))
train_labels = to_categorical(y_train, num_classes=40)
test_labels = to_categorical(y_test, num_classes=40)
X_train=np.dstack([X_train] * 3)
X_test=np.dstack([X_test]*3)
X_train = X_train.reshape(-1, 64,64,3)
X_test= X_test.reshape (-1,64,64,3)


In [23]:
from tensorflow.keras.applications import VGG16
input_layer=layers.Input(shape=(64,64,3))
model_vgg16=VGG16(weights='imagenet',input_tensor=input_layer,include_top=False)
last_layer=model_vgg16.output
flatten=layers.Flatten()(last_layer) 

# Add output layer with 40 neurons
output_layer=layers.Dense(40,activation='softmax')(flatten)
model=models.Model(inputs=input_layer,outputs=output_layer)
for layer in model.layers[:-1]:
    layer.trainable=False

In [24]:
model.summary()

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_4 (InputLayer)        [(None, 64, 64, 3)]       0         
                                                                 
 block1_conv1 (Conv2D)       (None, 64, 64, 64)        1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 64, 64, 64)        36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 32, 32, 64)        0         
                                                                 
 block2_conv1 (Conv2D)       (None, 32, 32, 128)       73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 32, 32, 128)       147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 16, 16, 128)       0   

In [25]:
model.compile(loss ='categorical_crossentropy', optimizer='adam', metrics =['accuracy',Precision(),Recall()])
model.fit (X_train, train_labels, epochs=5)

Epoch 1/5
10/10 [==============================] - 14s 1s/step - loss: 3.7200 - accuracy: 0.0688 - precision_3: 0.0000e+00 - recall_3: 0.0000e+00
Epoch 2/5
10/10 [==============================] - 14s 1s/step - loss: 2.9029 - accuracy: 0.4000 - precision_3: 0.0000e+00 - recall_3: 0.0000e+00
Epoch 3/5
10/10 [==============================] - 14s 1s/step - loss: 2.3863 - accuracy: 0.6469 - precision_3: 0.0000e+00 - recall_3: 0.0000e+00
Epoch 4/5
10/10 [==============================] - 12s 1s/step - loss: 1.9086 - accuracy: 0.8094 - precision_3: 1.0000 - recall_3: 0.0219
Epoch 5/5
10/10 [==============================] - 12s 1s/step - loss: 1.5541 - accuracy: 0.9062 - precision_3: 1.0000 - recall_3: 0.0656


In [26]:
print("Model evaluation on test data")
model.evaluate(X_test,test_labels)

Model evaluation on test data
3/3 [==============================] - 5s 1s/step - loss: 1.7012 - accuracy: 0.7750 - precision_3: 1.0000 - recall_3: 0.0875


[1.7011674642562866, 0.7749999761581421, 1.0, 0.08749999850988388]

We can see that the transfer learning technique gave us decent results (0.775 accuracy).

## Transfer learning - ResNet - faces dataset

In [33]:
from tensorflow import keras

X = faces['images']
x = keras.applications.resnet_v2.preprocess_input(tf.cast(X, tf.float32))
y = faces['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], X_train.shape[2], 1))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1], X_test.shape[2], 1))
train_labels = to_categorical(y_train, num_classes=40)
test_labels = to_categorical(y_test, num_classes=40)
x = tf.image.grayscale_to_rgb(tf.constant(X_train))

model = keras.Sequential()
base_model = keras.applications.ResNet152V2(
  include_top=False,
  weights='imagenet',
  input_shape=(64, 64, 3)
)
for layer in base_model.layers[:-2]:
    layer.trainable = False

model.add(base_model)
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(40, activation='softmax', kernel_regularizer=keras.regularizers.l2(0.001)))

In [34]:
X_train=np.dstack([X_train] * 3)
X_test=np.dstack([X_test]*3)
X_train = X_train.reshape(-1, 64,64,3)
X_test= X_test.reshape (-1,64,64,3)

In [35]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 resnet152v2 (Functional)    (None, 2, 2, 2048)        58331648  
                                                                 
 flatten_6 (Flatten)         (None, 8192)              0         
                                                                 
 dense_21 (Dense)            (None, 40)                327720    
                                                                 
Total params: 58,659,368
Trainable params: 331,816
Non-trainable params: 58,327,552
_________________________________________________________________


In [36]:
model.compile(loss ='categorical_crossentropy', optimizer='adam', metrics =['accuracy',Precision(),Recall()])
model.fit (X_train, train_labels, epochs=5)

Epoch 1/5
10/10 [==============================] - 19s 974ms/step - loss: 3.1506 - accuracy: 0.3031 - precision_6: 0.5897 - recall_6: 0.1437
Epoch 2/5
10/10 [==============================] - 10s 978ms/step - loss: 0.4202 - accuracy: 0.9469 - precision_6: 1.0000 - recall_6: 0.8500
Epoch 3/5
10/10 [==============================] - 10s 980ms/step - loss: 0.1465 - accuracy: 0.9969 - precision_6: 1.0000 - recall_6: 0.9844
Epoch 4/5
10/10 [==============================] - 10s 981ms/step - loss: 0.1103 - accuracy: 1.0000 - precision_6: 1.0000 - recall_6: 0.9937
Epoch 5/5
10/10 [==============================] - 10s 984ms/step - loss: 0.1019 - accuracy: 1.0000 - precision_6: 1.0000 - recall_6: 0.9969


In [37]:
print("Model evaluation on test data")
model.evaluate(X_test,test_labels)

Model evaluation on test data
3/3 [==============================] - 5s 758ms/step - loss: 1.5421 - accuracy: 0.7000 - precision_6: 0.7800 - recall_6: 0.4875


[1.5420591831207275,
 0.699999988079071,
 0.7799999713897705,
 0.48750001192092896]

#### Conclusion :  
VGG16 and Resnet50_v2 gave decent results in terms of accuracy on the test set. They also outperformed the hard-coded vgg16 architecture. These results highlight the importance of transfer learning since it helps us benefit of a complexe architecture which was trained on a very large dataset (ImageNet) with freezing the hidden layers and adapting the input/output layers to our data.